### DISCUSSION TITLE
Presentator v3 release and some real numbers regarding its PocketBase server setup

### DISCUSSION THREAD

**Discussion by ganigeorgiev** [pocketbase/pocketbase#4254] [Show and tell]
Earlier today I've released [Presentator v3](https://presentator.io/) (based on PocketBase) and I mentioned in the past discussions that will share some numbers, so here there are.

- [Server setup](#server-setup)
- [App details and numbers](#app-details-and-numbers)
- [Backups](#backups)
- [Custom backend code](#custom-backend-code)
- [What next](#what-next)

## Server setup

The application is currently hosted on the smallest Hetzner CAX11 instance (2vCPU Ampere/ARM64; 4GB RAM; 40GB disk space) and has an additional 210GB block storage volume mounted only for the uploaded files (`pb_data/storage`).

The total monthly cost would be **~€15**: €4.55 for the server and €11.09 for the extra block storage (_the prices are with 20% VAT so for your location may slightly vary; you can check your current detected location VAT rate from the header in the Hetzner website_).

Running for couple hours it seems to work fine so far without any noticible performance degradation (_in the worst case scenario I'll just rescale it from the Hetzner dashboard, but I doubt that it will be needed_).

## App details and numbers

The current Presentator v3 data is organized in 10 collections with various schema sizes and fields.

At the time or writing, the total number of records across all collections is **~573K**.
The largest collection has a little over **260k** records.
The "users" are **~17k** but I have no idea how much of those accounts are actually in use.

The database file (`pb_data/data.db`) currently is **~130MB**.

The uploaded files (`pb_data/storage`) are total of **~165GB**.

Note that this DB data and storage size is the accumulated amount over 6-7 years (_the first Presentator beta release was around 2015/2016 and for some time it was a private/invite only until open sourced and available for everyone couple months later_).

## Backups

For backups, I use an old spinning drive (500GB) connected via usb to my laptop that runs a bash script twice a day and once every Sunday.

I'm not sure if it would be useful but below is a simplified version of my backup script:
<details>
<summary><code>backup.sh</code></summary>

```sh
#!/bin/bash

VPS_IP="..."
VPS_PB_DATA="..."
LOCAL_PB_DATA="..."

# --------------------------------------------------------------------
# Simple shellscript to backup the remote PocketBase (data.db + storage)
# to a local directory.
#
# NB! Update the above variables according to your system.
#
# To add it as a cron task (crontab -e):
# ```sh
# # run on every 12h
# 0 */12 * * * /path/to/app/backup.sh
# ```
# --------------------------------------------------------------------

echo ""
echo "---------------------------------------------------------------"
echo "-> SQLite database backup"
echo "---------------------------------------------------------------"

# create a db backup file
ssh -T root@$VPS_IP << EOL
cd $VPS_PB_DATA
sqlite3 data.db '.backup data.db.backup'
exit
EOL

if [ $? -ne 0 ]; then
    exit 1
fi

# rsync the db backup file
rsync -av -e ssh root@$VPS_IP:$VPS_PB_DATA/data.db.backup $LOCAL_PB_DATA/data.db

if [ $? -ne 0 ]; then
    exit 1
fi

echo ""
echo "---------------------------------------------------------------"
echo "-> Storage files backup"
echo "---------------------------------------------------------------"

# rsync the storage files
rsync -av -e ssh root@$VPS_IP:$VPS_PB_DATA/storage/ $LOCAL_PB_DATA/storage/

if [ $? -ne 0 ]; then
    exit 1
fi

echo ""
echo "==============================================================="
echo "Backup completed!"
```
</details>

Basically it generates a backup of the SQLite db file and then calls `rsync` to incrementally sync the changes to my local usb disk directory (_note that we don't use the builtin PocketBase backup zip functionality because it creates a full archive of the `pb_data` and this will be very slow, even without compression, for 165GB+ many small files_).

## Custom backend code

Since we already had a lot of things builtin in PocketBase, there wasn't really that much extra backend code needed for Presentator.
You can check the repo if you are curious - https://github.com/presentator/presentator.
The custom backend functionality is mostly related to the app specific notifications and comments handling and most of the Go code in the repo is actually integration tests for the web APIs and collection rules. Everything else is frontend.

## What next

As most of you probably know, I started PocketBase because of Presentator, but that doesn't mean that its development will stall.

Quite the contrary actually. Now that most of the Presentator related work is pushed out of the way, I will be able to dedicate and focus most of my time on improving and fixing the PocketBase API warts.
And there are a lot of things that I want to change.
The current web APIs and exposed Go/JS interfaces are workable, but far from perfect.
There are a lot of ambiguities and caveats to the framework usage (especially related to the files upload handling) that can be done a lot better and simpler.

In other words, I want to make PocketBase a better, simpler and more extendable Go framework.

I'll take a short break the next couple days to eventually fix whatever show up related to the Presentator v3 release, but after that I'm planning to start designing and experimenting with a slightly different PocketBase internal structure. There will be more details about the planned changes once I test couple different ideas, but in general I'm planning to work towards making the SQLite dev experience more "native" (_ex. loading a raw SQLite db without manually mapping to the PocketBase fields_), allow custom user defined collection fields (initially only the Go code portion, the UI will be as final stage), simplify the models abstraction and make everything `Record`, eventually remove/refactor the form models to avoid having 2 different ways of persisting data, and many more. And again, I don't want to make promises and there are no fixed ETAs so please be understanding.

**Comment by katriellucas** [comment]
Can we know how many concurrent users Presentator v3 gets?

**Reply by ganigeorgiev** [reply]
I don't know because we don't track it.

If you are interested in the load/traffic that the app is handling, we can easily measure only the peak number of requests per hour (based on the PocketBase activity request logs, aka. `data.type="request"`). I'll wait to gather some more data and will report back tomorrow.

**Reply by ganigeorgiev** [reply]
After leaving it for a while the largest reqs/hour number seems to be 9593 (or ~160reqs/min), which is practically nothing.
The CPU usage based on the Hetzner dashboard graph for the last 24h remains under 20% for most of the time and memory usage seems to be between 200-600MB (_note that this most likely also include unused memory pages since the PocketBase process is the only major running program on the server and the OS may not reclaim the unused memory back right away as there is no memory pressure_).

For general PocketBase benchmarks performance of the used server you can also explore the [CAX11 instance results](https://github.com/pocketbase/benchmarks/blob/master/results/hetzner_cax11.md).

**Reply by katriellucas** [reply]
I see, thanks a lot for the stats. I just think that it's interesting. I'm yet to see someone pushing PB to it's limits or saying it has a bad performance in a production env. I'm highly considering PB for my next project and based on the tests alone.

**Comment by anbraten** [comment]
Thanks for the interesting insides.

For an alternative backupstorage just wanted to point you to https://www.hetzner.com/storage/storage-box/

**Comment by sujed** [comment]
Great job!! 👏👏
It would be cool to have some metrics from a production app if possible

**Comment by pnmcosta** [comment]
This is so encouraging. Thank you for sharing. Have an awesome break.

**Comment by orenaksakal** [comment]
@ganigeorgiev are you hosting the sveltekit on the same server? Or these stats are only for pocketbase? I'm wondering if I can host both on a cax11 and not worry about performance for a while until it's time to scale

**Reply by ganigeorgiev** [reply]
Everything is on the same server and embedded in the binary.

Presentator doesn't use JavaScript SSR and both the application UI (Svelte) and the landing site (SvelteKit) are static client-side apps.

**Comment by moussa32** [comment]
Thank you for your sharing

I am building a SaaS using Nextjs & Pocketbase and I was wondering about the performance in the production environment I hope you well and keep the good job 👍

**Comment by mkelepe** [comment]
@ganigeorgiev 

Regarding the backup script, is it okay copying the db files without first stopping pocketbase? Doesn't this create a risk of creating a corrupted backup? (I see that you rsync the db backup, not the db files, but I don't use pocketbase's built-in backup functionality because it block the writes to the db during the backup and I try to avoid that)

Ideally, I would like to achieve safe backups (avoid any risk of corruption) without blocking the writes on the db, so my users can continue using the app without interruptions (bonus: the backups to be incremental). Does rsync achieve this?

**Reply by ganigeorgiev** [reply]
It is safe because it is using the [`sqlite3 cli .backup`](https://www.sqlite.org/cli.html#special_commands_to_sqlite3_dot_commands_) dot-command (it is essentially the same as `VACUUM INTO`).

> without blocking the writes on the db...

There is no such option. If you want streaming replication you could check [LiteStream](https://litestream.io/) or other similar tools.
